lib/Dialect/Torch/Transforms/RefineTypes.cpp

```diff
@@ -1314,21 +1320,37 @@ ChangeResult TypeAnalyzer::visitAtenEmbeddingOp(
   return getLatticeElement(op.getResult()).join(knowledge);
 }
 
+template <typename OpTy>
+static ValueKnowledge
+getSameSizeAsInput(OpTy op,
+                   ArrayRef<LatticeElement<ValueKnowledge> *> operands) {
+  auto input = operands[0]->getValue();
+  auto knowledge =
+      ValueKnowledge::getNotNonePessimisticValueState(op->getContext());
+  knowledge.hasSizes = input.hasSizes;
+  knowledge.sizes = input.sizes;
+  return knowledge;
+}
 
 // Common template for softmax like ops, eg., log_softmax.
 template <typename OpTy>
 ChangeResult TypeAnalyzer::visitAtenSoftmaxLikeOp(
     OpTy op, ArrayRef<LatticeElement<ValueKnowledge> *> operands) {
   auto input = operands[0]->getValue();
   auto dtype = op.dtype();
-  auto knowledge =
-      ValueKnowledge::getNotNonePessimisticValueState(op->getContext());
-  knowledge.hasSizes = input.hasSizes;
-  knowledge.sizes = input.sizes;
+  ValueKnowledge knowledge = getSameSizeAsInput(op, operands);
   fillInDTypeGivenDTypeIntAndInputDType(knowledge, dtype, input.dtype);
   return getLatticeElement(op.getResult()).join(knowledge);
 }
 
+ChangeResult TypeAnalyzer::visitAten_SoftmaxOp(
+    Aten_SoftmaxOp op, ArrayRef<LatticeElement<ValueKnowledge> *> operands) {
+  auto input = operands[0]->getValue();
+  ValueKnowledge knowledge = getSameSizeAsInput(op, operands);
+  knowledge.dtype = input.dtype;
```

Comment:
I think you should bring back the logic you had for the `halfToFloat` type here. In the case of `DecomposeComplexOps`, we are avoiding dealing with `halfToFloat` because one has to be careful about changing element types correctly, but `RefineTypes` does not have to worry about any of this. `RefineTypes` just has to give an accurate return type.